src/include/txn.i

```diff
@@ -1005,7 +1006,14 @@ __wt_txn_read(WT_SESSION_IMPL *session, WT_CURSOR_BTREE *cbt, WT_ITEM *key, uint
         WT_ASSERT(session, F_ISSET(prepare_upd, WT_UPDATE_PREPARE_RESTORED_FROM_DS));
         if (prepare_upd->txnid == WT_TXN_ABORTED ||
           prepare_upd->prepare_state == WT_PREPARE_RESOLVED)
-            return (WT_RESTART);
+            /*
+             * When a prepared update/insert is rollback or committed, retrying it again should fix
+             * concurrent modification of a prepared update. Other than prepared insert rollback,
+             * rest of the cases, the history store update is either added to the end of the update
+             * chain or modified to set proper stop timestamp. In all the scenarios, retrying again
+             * will work to return a proper update.
+             */
+            goto retry;
```

Comment:
Could we have logic to restrict the retry with a counter? That will prevent a hang in case of other coding bugs as well ?

Reply:
The retry logic shouldn't enter twice for a single key. Instead of a counter, how about adding a flag and assert it to make sure that we didn't enter the retry logic twice?

Reply:
Yes, I understand that it is not supposed to enter more than once. My concern is to avoid hang, if there is some coding bug. Assert will not be available in release mode. not necessarily counter, but some logic to avoiding hang in both debug and release mode.

Reply:
Ok. I will add a bool flag before retrying the logic to ensure that we never entered twice so that it can work in release mode also.

Reply:
Added a flag and stats also to check how many times does this retry occur.